This adds a configuration option `boot_devices`

danielfdickinson opened the pull request:
to the domain config, which allows changing the boot device(s)
from the default of only from the first bootable hard drive ("hd").

I use this to set the boot device to "network" in
the plugin domain generation, and with a "managed" network which has:

``` xml
  <ip address='192.168.125.1' netmask='255.255.255.0'>
    <dhcp>
      <range start='192.168.125.128' end='192.168.125.254'/>
      <bootp file='http://ipxe-boot.example.net/boot.ipxe'/>
    </dhcp>
  </ip>
```

I am able to use iPXE to launch an instance. This is useful for
initial provisioning of Alpine Linux which does not have cloud-init
images for Qemu/KVM or Libvirt atthe present time.

The iPXE script I use is similar to:

---

```
set base-url http://ipxe-boot.example.com

kernel ${base-url}/boot-3.16/vmlinuz-virt console=tty0 modules=loop,squashfs quiet nomodeset alpine_repo=https://dl-cdn.alpinelinux.org/alpine/v3.16/main modloop=https://ipxe-boot.example.com/boot-3.16/modloop-virt ssh_key="ssh-ed25519 AAAA..."
initrd ${base-url}/boot-3.16/initramfs-virt
boot
```

---

where ssh_key is an actual SSH public key in authorized_keys format.

Signed-off-by: Daniel F. Dickinson <dfdpublic@wildtechgarden.ca>

thomasklein94 reviewed (changes requested):
Thank you for your contribution. I've found one major and one minor issue with your PR, please see the comments and suggestions.

danielfdickinson commented on builder/libvirt/config.go:
```diff
@@ -118,6 +122,10 @@ func (c *Config) Prepare(raws ...interface{}) ([]string, error) {
 		c.DomainName = fmt.Sprintf("packer-%s", xid.New())
 	}
 
+	if len(c.BootDevices) == 0 {
+		c.BootDevices[0] = "hd"
+	}
```
Thank you for the catch. It's been many years since I touched golang, and I think I need to spend some time relearning. This one is a bit annoying though as if I had done the testing I had intended (but apparently forgot), I would have caught this error.
Making the change and doing doing more complete testing, and then will update.

thomasklein94 replied:
Thank you.
About the testing part, I'm also guilty as I've not yet included any automated testing in the pipeline. That's something I should put on my roadmap soon. But meanwhile, I appreciate you testing this changeset manually.

danielfdickinson commented on builder/libvirt/domain.go:
```diff
@@ -64,6 +65,12 @@ func newDomainDefinition(config *Config) libvirtxml.Domain {
 		},
 	}
 
+	for _, bd := range config.BootDevices {
+		bootDevice := &libvirtxml.DomainBootDevice{}
+		bootDevice.Dev = bd
+		domainDef.OS.BootDevices = append(domainDef.OS.BootDevices, *bootDevice)
+	}
```
Thank you again. I confess to a little too much copypasta hackery on this one. It worked, but definitely appreciate the improved version. Testing again, and then will update.

danielfdickinson commented:
I've fixed the issues and added proper validation of the boot devices (though perhaps that should moved to a separate function?) and tested with no `boot_devices`, with a single boot devices, with multiple valid boot devices, and with various combinations of valid and invalid boot devices.

thomasklein94 commented:
Your changes are released in [v0.3.1](https://github.com/thomasklein94/packer-plugin-libvirt/releases/tag/v0.3.1). 
Thank you for your contribution
